### PR TITLE
SWT reproducibility

### PR DESCRIPTION
**Pull request by bb010g**
###### Motivation for this change

More reproducibility! And less likelihood of breakage moving forward into new versions. Also, the build-support/setup-hooks/canonicalize-jars-hook hook added to support this was pulled from the unused build-support/release/ant-build (& functions only used there sourced from build-support/release/functions.sh), which has been deleted due to being weird and old and not imported by any package built on Hydra in the last 3+ years.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
  - Tested with applications/editor/music/tuxguitar & applications/networking/ipscan (in [my NUR repo](https://github.com/bb010g/nur-packages/blob/a9b04ae395780ab23d09b8657515436ceb8c97e4/pkgs/applications/networking/ipscan/default.nix); planning to submit to Nixpkgs after this merges since it requires upgraded SWT).
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - **Decrease** from 648420920 to 648403736
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bb010g** [comment]
@edolstra may also want to review this, as I think the `build-support/release` stuff was mostly him?

File: pkgs/build-support/java/canonicalize-jar.sh
Comment (rycee) [inline pkgs/build-support/java/canonicalize-jar.sh]: Could use the `-m` option and have
```sh
@zip@ -qroXm "$outer/tmp-out.jar" "$input" 2> /dev/null
```
perhaps? Then could remove the `mv "$outer/tmp-out.jar" "$input"` below and change `rm -rf "$input-tmp"` to `rmdir "$input-tmp"`.

File: pkgs/development/libraries/java/swt/default.nix
Comment (rycee) [inline pkgs/development/libraries/java/swt/default.nix]:
```
description = "A widget toolkit for Java";
longDescription = ''
  This widget toolkit allows Java to access the user-interface
  facilities of the operating systems on which it is implemented.
'';
```

**Review by rycee — COMMENTED** [review]
I think this is very nice! Having the ability to easily canonicalize Jar files is very convenient!

File: pkgs/build-support/java/canonicalize-jar.sh
Comment (bb010g) [reply pkgs/build-support/java/canonicalize-jar.sh]: This won't work because then you'll record the input directory name. We need to be inside `$input-tmp".

**Comment by bb010g** [comment]
Ok, rebased and ready for review.

**Comment by doronbehar** [comment]
Please fix the merge conflicts.

**Comment by bb010g** [comment]
Rebased.

**Comment by bb010g** [comment]
Rebased.

**Comment by bb010g** [comment]
Rebased. Reminder that this is blocking SWT upgrades.

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18718]: Those need alias entries.

File: pkgs/development/libraries/java/swt/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/java/swt/default.nix:123]: Please undo these changes.

File: pkgs/development/libraries/java/swt/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/java/swt/default.nix:115]:
```suggestion
    homepage = "https://www.eclipse.org/swt/";
```

File: pkgs/development/libraries/java/swt/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/java/swt/default.nix:99]:
```suggestion
    bash ./build.sh
```

File: pkgs/development/libraries/java/swt/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/java/swt/default.nix:94]: This monster needs more comments explaining what is done.

File: pkgs/development/libraries/java/swt/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/java/swt/default.nix:33]:
```suggestion
    url = "https://archive.eclipse.org/eclipse/downloads/drops4/R-${fullVersion}/${pname}-${version}-${metadata.platform}.zip";
```

File: pkgs/build-support/setup-hooks/canonicalize-jars.sh
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/canonicalize-jars.sh:17]:
```suggestion
source @canonicalize_jar@
```

File: pkgs/build-support/java/canonicalize-jar.sh
Comment (SuperSandro2000) [inline pkgs/build-support/java/canonicalize-jar.sh:28]:
```suggestion
    (head -n 1 "$input" && tail -n +2 "$input" | sort | grep -v '^\s*$') > "$input-tmp"
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by fzakaria** [comment]
I recently wrote https://fzakaria.com/2021/06/27/java-nix-reproducibility.html
Curious about the delta between the hook and other canonicalization vs. just adding _strip-nondeterminism_ to the fixupPhase?

**Comment by fzakaria** [comment]
Can the PR description's motivation be augmented to describe the problem and the fixes?
I'm aware of the mtime fixes but not the other canonicalizations that are needed.

Doing it in a hook vs the fixupPhase is also of interest (see earlier comment)

File: pkgs/top-level/all-packages.nix
Comment (bb010g) [reply pkgs/top-level/all-packages.nix]: `junixsocket` depends on `junit`. `junit` has been marked as broken since 2015-09-08, i.e. 6 years ago. Nobody has submitted a patch for `junit` during the past 6 years. Any downstream who currently unbreaks `junit` will be fine vendoring it; those downstreams basically have been vendoring `junit` for the past 6 years.

This lack of need of aliases is now better explained in the message of the commit that removes them.

**Comment by bb010g** [comment]
Rebased, and @pSub removed as maintainer due to a lack of activity with this PR over the last 2 years.

@fzakaria It looks like `META-INF/MANIFEST.MF` is sorted & empty lines removed? It's been a while since I submitted this patch; that behavior was carried over from the original `canonicalizeJarManifest` function. Further documentation or fixes are welcome; I know that this PR works and just want it to finally get merged so I can use `ipscan` from upstream Nixpkgs.

> Doing it in a hook vs the fixupPhase is also of interest

`canonicalizeJarsIn` is called in `fixupOutputHooks`, which are ran directly after `runHook preFixup` in the default `fixupPhase`.

File: pkgs/top-level/all-packages.nix
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:154]: Maybe the commit that adds these hook and script should use them in `pkgs/build-support/release/functions.sh` somehow? I dislike that we have such code duplication.

File: pkgs/development/libraries/java/swt/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/java/swt/default.nix]: Why is this needed? Maybe it should be commented?

File: pkgs/development/libraries/java/swt/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/java/swt/default.nix:73]: Have you considered applying a patch and suggesting it upstream?

**Review by doronbehar — COMMENTED** [review]
I'd also squash the last two commits into 1.

File: pkgs/top-level/all-packages.nix
Comment (bb010g) [reply pkgs/top-level/all-packages.nix:154]: I disliked it too, but the functions they're duplicated from are removed next commit, and it seemed like busywork.

These commits were originally merged, but splitting them up allowed the commit messages to get clearer.

File: pkgs/development/libraries/java/swt/default.nix
Comment (bb010g) [reply pkgs/development/libraries/java/swt/default.nix]: This `extraPostFetch` performs the second unpack necessary.

File: pkgs/development/libraries/java/swt/default.nix
Comment (bb010g) [reply pkgs/development/libraries/java/swt/default.nix:73]: Yes, but I figured upstream submissions could be figured out after our SWT was up-to-date first. I don't think 4.5 is still maintained. My original idea was to upgrade SWT after this was merged first, to keep each PR easier to follow.

File: pkgs/development/libraries/java/swt/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/java/swt/default.nix]: > This `extraPostFetch` performs the second unpack necessary.

You mean upstream sends us a zip inside a zip?

File: pkgs/development/libraries/java/swt/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/java/swt/default.nix:73]: I tried to update SWT once..

https://github.com/NixOS/nixpkgs/pull/75609

File: pkgs/development/libraries/java/swt/default.nix
Comment (bb010g) [reply pkgs/development/libraries/java/swt/default.nix]: Yeah. Per the comment above:

```nix
  # Alas, the Eclipse Project apparently doesn't produce source-only
  # releases of SWT.  So we just grab a binary release and extract
  # "src.zip" from that.
```

File: pkgs/development/libraries/java/swt/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/java/swt/default.nix:139]:
```suggestion
      A widget toolkit for Java to access the user-interface facilities of
```

File: pkgs/development/libraries/java/swt/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/java/swt/default.nix:51]: Why do you need to use `pkgsBuildHost`?

**Review by doronbehar — COMMENTED** [review]
I noticed some dependencies can be cleaned and added:

```diff
diff --git i/pkgs/development/libraries/java/swt/default.nix w/pkgs/development/libraries/java/swt/default.nix
index 348af9d255d..f965a1dc3c5 100644
--- i/pkgs/development/libraries/java/swt/default.nix
+++ w/pkgs/development/libraries/java/swt/default.nix
@@ -1,6 +1,15 @@
-{ lib, stdenv, fetchzip, pkgsBuildHost
-, atk, glib, gtk2, jdk, libGL, libGLU, libXi, libXt, libXtst, libsoup
-, webkitgtk, xorg
+{ lib
+, stdenv
+, fetchzip
+, pkgsBuildHost
+, atk
+, glib
+, gtk2
+, jdk
+, libGL
+, libGLU
+, libXtst
+, gnome2
 }:
 
 let
@@ -56,11 +65,10 @@ in stdenv.mkDerivation rec {
     jdk
     libGL
     libGLU
-    libXi
-    libXt
     libXtst
-    libsoup
-    webkitgtk
+    gnome2.gnome_vfs
+    gnome2.libgnomeui
+    gnome2.libgnome
   ];
 
   patches = [ ./awt-libs.patch ./gtk-libs.patch ];
```

File: pkgs/development/libraries/java/swt/default.nix
Comment (kylesferrazza) [reply pkgs/development/libraries/java/swt/default.nix:85]: Do you still plan to update SWT now that this PR has been merged? @bb010g

**Pull request merged**
